src/cluster_migrate.c

```diff
@@ -0,0 +1,1986 @@
+                addReplyErrorObject(c, shared.syntaxerr);
+                goto cleanup;
+            }
+            source_node = clusterLookupNode(c->argv[4]->ptr, CLUSTER_NAMELEN);
+            if (!source_node) {
+                addReplyError(c, "Target node does not know the source node");
+                goto cleanup;
+            }
+            i += 2;
+            continue;
+        }
+        if (!strcasecmp(c->argv[i]->ptr, "linkname")) {
+            if (link_name || i + 1 >= c->argc || sdslen(c->argv[i + 1]->ptr) != CLUSTER_NAMELEN) {
+                addReplyErrorObject(c, shared.syntaxerr);
+                goto cleanup;
+            }
+            link_name = c->argv[i + 1]->ptr;
+            i += 2;
+            continue;
+        }
+        if (!strcasecmp(c->argv[i]->ptr, "slotsrange")) {
+            if (slot_ranges) {
+                addReplyErrorObject(c, shared.syntaxerr);
+                goto cleanup;
+            }
+            /* parseSlotRanges will set i for the next iteration */
+            slot_ranges = parseSlotRangesOrReply(c, i + 1, &i, &owning_node);
+            if (slot_ranges == NULL) {
+                goto cleanup;
+            }
+
+            listIter li;
+            listNode *ln;
+            listRewind(slot_ranges, &li);
+            while ((ln = listNext(&li)) != NULL) {
+                slotRange *range = ln->value;
+                for (int i = range->start_slot; i <= range->end_slot; i++) {
+                    if (clusterIsSlotImporting(i)) {
+                        addReplyError(c, "Slot is already being imported on the target by a different migration");
+                        goto cleanup;
+                    }
+                }
+            }
+
+            continue;
+        }
+        addReplyErrorObject(c, shared.syntaxerr);
+        goto cleanup;
+    }
+    if (!source_node || !link_name || !slot_ranges) {
+        addReplyErrorObject(c, shared.syntaxerr);
+        goto cleanup;
+    }
+    if (source_node != owning_node) {
+        addReplyError(c, "Target node does not agree about current slot ownership");
+        goto cleanup;
+    }
+
+    slotMigrationLink *link = createSlotImportLink(c, source_node, link_name, slot_ranges);
+    listAddNodeHead(server.cluster->slot_migration_links, link);
+
+    clusterDoBeforeSleep(CLUSTER_TODO_HANDLE_SLOT_MIGRATION);
+
+    /* Push the reply, which will allow the migration to continue */
+    ClientFlags old_flags = link->client->flag;
+    link->client->flag.pushing = 1;
+    addReply(c, shared.ok);
+    link->client->flag = old_flags;
+    return;
+
+cleanup:
+    if (slot_ranges) {
+        listRelease(slot_ranges);
+    }
+    return;
+}
+
+/* Sent by the source to the target after dumping the snapshot in AOF format. */
+void clusterCommandSyncSlotsSnapshotEof(client *c) {
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
```

Comment:
```suggestion
    slotMigrationLink *link = c->slot_migration_link;
```
You should be able to forward declare the type so that we can avoid all this casting. It seems to be done a lot.

Reply:
Makes sense. Done.